module/support/logging/DataLogging/src/DataLogging.cpp

```diff
@@ -72,61 +74,22 @@ namespace module::support::logging {
+                // Creates the output ".nbs" file path.
                 output_file_path = std::filesystem::path();
                 output_file_path += config.output.directory / config.output.binary / ("_" + formatted_time() + ".nbs");
-                output_file = std::ofstream(output_file_path);
 
-                // Creates the output ".idx" file.
+                // Creates the output ".idx" file path.
                 index_file_path = output_file_path;
                 index_file_path += ".idx";
-                index_file = std::make_unique<zstr::ofstream>(index_file_path);
 
-                bytes_written = 0;
+                encoder = std::make_unique<utility::nbs::Encoder>(output_file_path, index_file_path);
             }
 
-            // Write radiation symbol
-            output_file.put(char(0xE2));
-            output_file.put(char(0x98));
-            output_file.put(char(0xA2));
-
-            // Write the size of the packet
-            output_file.write(reinterpret_cast<const char*>(&size), sizeof(size));
-
-            // Write the timestamp
-            output_file.write(reinterpret_cast<const char*>(&timestamp_us), sizeof(timestamp_us));
-
-            // Write the hash
-            output_file.write(reinterpret_cast<const char*>(&data.hash), sizeof(data.hash));
-
-            // Write the actual packet data
-            output_file.write(data.data.data(), data.data.size());
-            output_file.flush();
-
-            // NBS Index File Format
-            // Name      | Type               |  Description
-            // ------------------------------------------------------------
-            // hash      | uint64_t           | the 64bit hash for the payload type
-            // id        | uint32_t           | the id field of the payload (or 0 if it does not have one)
-            // timestamp | uint64_t           | Timestamp of the message or the emit timestamp in nanoseconds
-            // offset    | uint64_t           | offset to start of radiation symbol ☢
-            // size      | uint32_t           | Size of the whole packet from the radiation symbol
-
-            uint32_t full_size = size + 7;
-
-            index_file->write(reinterpret_cast<const char*>(&data.hash), sizeof(data.hash));
-            index_file->write(reinterpret_cast<const char*>(&data.id), sizeof(data.id));
-            index_file->write(reinterpret_cast<const char*>(&data.message_timestamp), sizeof(data.message_timestamp));
-            index_file->write(reinterpret_cast<const char*>(&bytes_written), sizeof(bytes_written));
-            index_file->write(reinterpret_cast<const char*>(&full_size), sizeof(full_size));
-            index_file->flush();
-
-            // Update the number of bytes we have written to the nbs file
-            bytes_written += size + 3 + sizeof(timestamp_us) + sizeof(data.hash);
+            encoder->write(data.timestamp, data.message_timestamp, data.hash, data.id, data.data);
```

Comment:
```suggestion
            if (encoder) {
                encoder->write(data.timestamp, data.message_timestamp, data.hash, data.id, data.data);
            }
```

Reply:
Here `encoder` should be non-null, as in the scope above we construct it.